lib/Support/Unix/Program.inc

```diff
@@ -434,13 +434,24 @@ llvm::sys::writeFileWithEncoding(StringRef FileName, StringRef Contents,
 bool llvm::sys::commandLineFitsWithinSystemLimits(StringRef Program,
                                                   ArrayRef<const char *> Args) {
   static long ArgMax = sysconf(_SC_ARG_MAX);
+  // POSIX requires that _POSIX_ARG_MAX is 4096, which is the lowest possible
+  // value for ARG_MAX on a POSIX compliant system.
+  static long ArgMin = _POSIX_ARG_MAX;
+
+  // This the same baseline used by xargs.
+  long EffectiveArgMax = 128 * 1024;
+
+  if (EffectiveArgMax > ArgMax)
+    EffectiveArgMax = ArgMax;
+  else if (EffectiveArgMax < ArgMin)
+    EffectiveArgMax = ArgMin;
 
   // System says no practical limit.
   if (ArgMax == -1)
     return true;
 
   // Conservatively account for space required by environment variables.
-  long HalfArgMax = ArgMax / 2;
+  long HalfArgMax = EffectiveArgMax / 2;
```

Comment:
The code from lines 453 on is a bit tricky to parse and would benefit IMO from being factored out, with a nice Doxygen comment, basically the comment on line 460. Call the factored-out routine something like: `isAnyArgumentTooBig`. Then line 450 becomes something like: `return EffectiveArgLengthLimit == NoPracticalLimit || !isAnyArgumentTooBig(...)`. Clearer, IMO.